src/uct/api/uct.h

```diff
@@ -2063,25 +2064,136 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
 ucs_status_t uct_md_query(uct_md_h md, uct_md_attr_t *md_attr);
 
 
+/**
+ * @ingroup UCT_MD
+ * @brief UCT allocation parameters specification field mask
+ *
+ * The enumeration allows specifying which fields in @ref uct_mem_alloc_params_t
+ * are present.
+ */
+typedef enum {
+    /** Enables @ref uct_mem_alloc_params_t::flags */
+    UCT_MEM_ALLOC_PARAM_FIELD_FLAGS          = UCS_BIT(0),
+
+    /** Enables @ref uct_mem_alloc_params_t::address_p */
+    UCT_MEM_ALLOC_PARAM_FIELD_ADDR_PTR       = UCS_BIT(1),
+
+    /** Enables @ref uct_mem_alloc_params_t::length_p */
+    UCT_MEM_ALLOC_PARAM_FIELD_LENGTH_PTR     = UCS_BIT(2),
+
+    /** Enables @ref uct_mem_alloc_params_t::mem_type */
+    UCT_MEM_ALLOC_PARAM_FIELD_MEM_TYPE       = UCS_BIT(3),
+
+    /** Enables @ref uct_mem_alloc_params_t::methods */
+    UCT_MEM_ALLOC_PARAM_FIELD_METHODS        = UCS_BIT(4),
+
+    /** Enables @ref uct_mem_alloc_params_t::mds */
+    UCT_MEM_ALLOC_PARAM_FIELD_MDS            = UCS_BIT(5),
+
+    /** Enables @ref uct_mem_alloc_params_t::name */
+    UCT_MEM_ALLOC_PARAM_FIELD_NAME           = UCS_BIT(6)
+} uct_mem_alloc_params_field_t;
+
+
+/**
+ * @ingroup UCT_MD
+ * @brief Parameters for allocating memory using @ref uct_mem_alloc,
+ *        @ref uct_md_mem_alloc
+ */
+typedef struct {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_mem_alloc_params_field_t. Fields not specified in this mask will
+     * be ignored
+     */
+    uint64_t           field_mask;
+
+    /**
+     * Memory allocation flags, see @ref uct_md_mem_flags
+     */
+    unsigned           flags;
+
+    /**
+     * If @a address_p content is NULL, the underlying allocation routine will
+     * choose the address at which to create the mapping. If @a address_p
+     * content is non-NULL but UCT_MD_MEM_FLAG_FIXED is not set, the address
+     * will be interpreted as a hint as to where to establish the mapping. If
+     * @a address_p is non-NULL and UCT_MD_MEM_FLAG_FIXED is set, then the
+     * specified address is interpreted as a requirement. In this case, if the
+     * mapping to the exact address cannot be made, the allocation request
+     * fails.
+     */
+    void               **address_p;
+
+    /**
+     * Points to minimal size to allocate. The actual size may be larger, for
+     * example because of alignment restrictions. Upon successful return,
+     * filled with the actual size that was allocated. Must be >0.
+     */
+    size_t             *length_p;
```

Comment:
uct_mem_alloc already returns uct_allocated_memory_t as "out" parameter, why do we need to also pass address and length by pointer?